packages/schema-extract/src/file-linker.ts

```diff
@@ -0,0 +1,192 @@
+import * as ts from 'typescript';
+import { transform } from '../src/file-transformer';
+import { Schema, IObjectFields, ClassSchemaId, ClassSchema, ModuleSchema, isRef, isSchemaOfType, isClassSchema } from './json-schema-types';
+
+export class SchemaLinker {
+    private checker: ts.TypeChecker;
+    private program: ts.Program;
+
+    constructor(program: ts.Program, checker: ts.TypeChecker) {
+        this.checker = checker;
+        this.program = program;
+    }
+
+    public flatten(file: string, name: string, moduleId: string): Schema {
+        const schema = transform(this.checker, this.program.getSourceFile(file)!, '/src/' + moduleId, '/someProject');
+        if (!schema.definitions) {
+            return {};
+        }
+        const entity = schema.definitions[name];
+        if (!entity) {
+            return {};
+        }
+        return link(entity, schema);
+    }
+}
+
+function link(entity: Schema, schema: ModuleSchema): Schema {
+    if (isClassSchema(entity)) {
+        return linkClass(schema, entity);
+    }
+    if (isRef(entity)) {
+        const entityType = entity.$ref.replace('#', '');
+        const refEntity = schema.definitions![entityType];
+        if (!refEntity) {
+            return entity;
+        }
+        if (!refEntity.genericParams || !entity.genericArguments) {
+            return refEntity;
+        }
+        if (isSchemaOfType('object', refEntity)) {
+            const pMap = new Map();
+            refEntity.genericParams!.forEach((param, index) => {
+                pMap.set(`#${entityType}!${param.name}`, entity.genericArguments![index]);
+            });
+            return linkObject(entity, entityType, refEntity, pMap, schema);
+        } else {
+            return entity;
+        }
+    }
+    if (entity.$allOf) {
+        const res = handleIntersection(entity.$allOf, schema);
+        res.type = 'object';
+        return res;
+    }
+    if (entity.$oneOf) {
+        const res: Schema = {type: 'object', $oneOf: []};
+        for (const type of entity.$oneOf) {
+            res.$oneOf!.push(link(type, schema));
+        }
+        return res;
+    }
+    return entity;
+}
+
+function handleIntersection(options: Schema[], schema: ModuleSchema, paramsMap?: Map<string, Schema>): Schema {
+    const res: Schema & IObjectFields = {properties: {}};
+    for (const option of options) {
+        if (isRef(option)) {
+            let entity: Schema & IObjectFields;
+            if (paramsMap) {
+                entity = paramsMap!.get(option.$ref)!;
+                if (!entity) {
+                    return res;
+                }
+            } else {
+                const refEntity = option.genericArguments ? option : schema.definitions![option.$ref!.replace('#', '')];
+                entity = link(refEntity, schema);
```

Comment:
if only one of them has it, the result does not
if both have it, , the result should have an index signature thats the intersection of both.

its worth to note that index signature for known keys i.e.

type A = 'a' | 'b'
type B = 'a' | 'c'
type A1 = {
    [key:string]:string
}
type B1 = {
    [key in B]:"gaga"
}
type c = A1 & B1;

should be linked so that c is treated as

type c = {
    a:"gaga",
    b:string,
    c:"gaga"
}